src/measure_theory/lp_space.lean

```diff
@@ -276,52 +371,127 @@ begin
-  snorm f p μ ≤ snorm f q μ :=
+  snorm' f p μ ≤ snorm' f q μ :=
 begin
-  have h_le_μ := snorm_le_snorm_mul_rpow_measure_univ hp0_lt hpq μ hf,
+  have h_le_μ := snorm'_le_snorm'_mul_rpow_measure_univ hp0_lt hpq hf,
   rwa [measure_univ, ennreal.one_rpow, mul_one] at h_le_μ,
 end
 
-lemma mem_ℒp.mem_ℒp_of_exponent_le {p q : ℝ} {μ : measure α} [finite_measure μ] {f : α → E}
-  (hfq : mem_ℒp f q μ) (hp_pos : 0 < p) (hpq : p ≤ q) :
-  mem_ℒp f p μ :=
+lemma snorm'_le_snorm_inf (hp_pos : 0 < p) {f : α → F} [probability_measure μ] :
+  snorm' f p μ ≤ snorm_inf f μ :=
 begin
-  cases hfq with hfq_m hfq_lt_top,
-  split,
-  { exact hfq_m, },
-  have hq_pos : 0 < q, from lt_of_lt_of_le  hp_pos hpq,
-  suffices h_snorm : snorm f p μ < ⊤,
-  { have h_top_eq : (⊤ : ennreal) = ⊤ ^ (1/p), by simp [hp_pos],
-    rw [snorm, h_top_eq] at h_snorm,
-    have h_snorm_pow : ((∫⁻ (a : α), ↑(nnnorm (f a)) ^ p ∂μ) ^ (1/p)) ^ p < (⊤ ^ (1/p)) ^ p,
-    from ennreal.rpow_lt_rpow h_snorm hp_pos,
-    rw [←ennreal.rpow_mul, ←ennreal.rpow_mul] at h_snorm_pow,
-    simpa [(ne_of_lt hp_pos).symm] using h_snorm_pow, },
-  calc snorm f p μ
-      ≤ snorm f q μ * (μ set.univ) ^ (1/p - 1/q) :
-    snorm_le_snorm_mul_rpow_measure_univ hp_pos hpq μ hfq_m
+  refine le_trans (snorm'_le_snorm_inf_mul_rpow_measure_univ hp_pos) (le_of_eq _),
+  simp [measure_univ],
+end
+
+lemma snorm_le_snorm_of_exponent_le {p q : ennreal} (hpq : p ≤ q) [probability_measure μ]
+  {f : α → E} (hf : ae_measurable f μ) :
+  snorm f p μ ≤ snorm f q μ :=
+begin
+  by_cases hp0 : p = 0,
+  { simp [hp0], },
+  rw ←ne.def at hp0,
+  by_cases hq_top : q = ⊤,
+  { by_cases hp_top : p = ⊤,
+    { rw [hq_top, hp_top],
+      exact le_refl _, },
+    { have hp_pos : 0 < p.to_real,
+      from ennreal.to_real_pos_iff.mpr ⟨lt_of_le_of_ne (zero_le _) hp0.symm, hp_top⟩,
+      rw [snorm_eq_snorm' hp0 hp_top, hq_top, snorm_eq_snorm_inf],
+      refine le_trans (snorm'_le_snorm_inf_mul_rpow_measure_univ hp_pos) (le_of_eq _),
+      simp [measure_univ], }, },
+  { have hp_top : p ≠ ⊤,
+    { by_contra hp_eq_top,
+      push_neg at hp_eq_top,
+      refine hq_top _,
+      rwa [hp_eq_top, top_le_iff] at hpq, },
+    have hp_pos : 0 < p.to_real,
+    from ennreal.to_real_pos_iff.mpr ⟨lt_of_le_of_ne (zero_le _) hp0.symm, hp_top⟩,
+    have hq0 : q ≠ 0,
+    { by_contra hq_eq_zero,
+      push_neg at hq_eq_zero,
+      have hp_eq_zero : p = 0, from le_antisymm (by rwa hq_eq_zero at hpq) (zero_le _),
+      rw [hp_eq_zero, ennreal.zero_to_real] at hp_pos,
+      exact (lt_irrefl _) hp_pos, },
+    have hpq_real : p.to_real ≤ q.to_real, by rwa ennreal.to_real_le_to_real hp_top hq_top,
+    rw [snorm_eq_snorm' hp0 hp_top, snorm_eq_snorm' hq0 hq_top],
+    exact snorm'_le_snorm'_of_exponent_le hp_pos hpq_real _ hf, },
+end
+
+lemma snorm'_lt_top_of_snorm'_lt_top_of_exponent_le {p q : ℝ} [finite_measure μ] {f : α → E}
+  (hf : ae_measurable f μ) (hfq_lt_top : snorm' f q μ < ⊤) (hp_pos : 0 < p) (hpq : p ≤ q) :
```

Comment:
Do you mean that I can relax it to `0 <= p` (which is indeed easy), or do you see a way to have it for all p in R ?
Since snorm' is not used for `p <= 0`, I am reluctant to work a lot on the negative case.

Reply:
I meant `0 <= p`.